polygon/bridge/snapshot_store.go

```diff
@@ -0,0 +1,397 @@
+	var lastSegment *snapshotsync.VisibleSegment
+	for i := len(segments) - 1; i >= 0; i-- {
+		if segments[i].Src().Index() != nil {
+			gg := segments[i].Src().MakeGetter()
+			if gg.HasNext() {
+				lastSegment = segments[i]
+				break
+			}
+		}
+	}
+	if lastSegment == nil {
+		return 0
+	}
+	var lastBlockNum uint64
+	var buf []byte
+	gg := lastSegment.Src().MakeGetter()
+	for gg.HasNext() {
+		buf, _ = gg.Next(buf[:0])
+		lastBlockNum = binary.BigEndian.Uint64(buf[length.Hash : length.Hash+length.BlockNum])
+	}
+
+	return lastBlockNum
+}
+
+func (s *snapshotStore) LastProcessedBlockInfo(ctx context.Context) (ProcessedBlockInfo, bool, error) {
+	if blockInfo, ok, err := s.Store.LastProcessedBlockInfo(ctx); ok {
+		return blockInfo, ok, err
+	}
+
+	tx := s.snapshots.ViewType(heimdall.Events)
+	defer tx.Close()
+	segments := tx.VisibleSegments
+
+	if len(segments) == 0 {
+		return ProcessedBlockInfo{}, false, nil
+	}
+
+	if s.sprintLengthCalculator == nil {
+		return ProcessedBlockInfo{}, false, fmt.Errorf("can't calculate last block: missing sprint length calculator")
+	}
+
+	lastBlockNum := segments[len(segments)-1].To() - 1
+	sprintLen := s.sprintLengthCalculator.CalculateSprintLength(lastBlockNum)
+	lastBlockNum = (lastBlockNum / sprintLen) * sprintLen
+
+	return ProcessedBlockInfo{
+		BlockNum: lastBlockNum,
+	}, true, nil
+}
+
+func (s *snapshotStore) LastEventId(ctx context.Context) (uint64, error) {
+	lastEventId, err := s.Store.LastEventId(ctx)
+
+	if err != nil {
+		return 0, err
+	}
+
+	snapshotLastEventId := s.LastFrozenEventId()
+	if snapshotLastEventId > lastEventId {
+		return snapshotLastEventId, nil
+	}
+
+	return lastEventId, nil
+}
+
+func (s *snapshotStore) LastFrozenEventId() uint64 {
+	if s.snapshots == nil {
+		return 0
+	}
+
+	tx := s.snapshots.ViewType(heimdall.Events)
+	defer tx.Close()
+	segments := tx.VisibleSegments
+
+	if len(segments) == 0 {
+		return 0
+	}
+	// find the last segment which has a built index
+	var lastSegment *snapshotsync.VisibleSegment
+	for i := len(segments) - 1; i >= 0; i-- {
+		if segments[i].Src().Index() != nil {
+			gg := segments[i].Src().MakeGetter()
+			if gg.HasNext() {
+				lastSegment = segments[i]
+				break
+			}
+		}
+	}
+	if lastSegment == nil {
+		return 0
+	}
+	var lastEventId uint64
+	gg := lastSegment.Src().MakeGetter()
+	var buf []byte
+	for gg.HasNext() {
+		buf, _ = gg.Next(buf[:0])
+		lastEventId = binary.BigEndian.Uint64(buf[length.Hash+length.BlockNum : length.Hash+length.BlockNum+8])
+	}
+	return lastEventId
+}
+
+func (s *snapshotStore) LastProcessedEventId(ctx context.Context) (uint64, error) {
+	lastEventId, err := s.Store.LastProcessedEventId(ctx)
+
+	if err != nil {
+		return 0, err
+	}
+
+	snapshotLastEventId := s.LastFrozenEventId()
+	if snapshotLastEventId > lastEventId {
+		return snapshotLastEventId, nil
+	}
+
+	return lastEventId, nil
+}
+
+func (r *snapshotStore) EventLookup(ctx context.Context, txnHash libcommon.Hash) (uint64, bool, error) {
```

Comment:
There's a few more like this below it.

```suggestion
func (s *snapshotStore) EventLookup(ctx context.Context, txnHash libcommon.Hash) (uint64, bool, error) {
```